Support timed events for dtrace + other improvements

skywhale opened the pull request:
* Send multiple, timestamped events in `EventMessage`
* Add a proper ctrl-c handling to prevent kernel crash in rare circumstances
* Make the process ID argument an optional, to probe the entire kernel

goodhoko commented on crates/dtrace_probe/Cargo.toml:
```diff
@@ -7,5 +7,6 @@ edition = "2021"
 color-eyre = "0.6"
 clap = { version = "4.2", features = ["derive"] }
 composer_api = { path = "../composer_api" }
+ctrlc = "3.2"
```
I suppose we're using this library so that we can clean-up properly on exit by calling `dtrace.stop()`, right? Then we might consider this
```suggestion
ctrlc = { version = "3.2", features = ["termination"] }
```
which makes ctrlc handle the SIGTERM and SIGHUP signas as well.

skywhale replied:
A good trick. Thank you :)

goodhoko commented on crates/dtrace_probe/src/main.rs:
```diff
@@ -28,19 +44,22 @@ fn main() -> Result<()> {
         for probe in &result.probes {
-            // TODO(skywhale): Use this timestamp.
-            let _timestamp: u128 = probe.traces[0].parse()?;
+            let timestamp = Duration::from_nanos(probe.traces[0].parse()?);
```
I'm thinking if we should rather panic on failing `parse`. The result is practically the same - the program exits. Conceptually, not being able to parse the timestamp is non-recoverable. Something's really wrong if what we get here is not a number.

We *do* panic when `probe.traces[0]` fail and that's the same severity. I think we should either panic or early-return error in both cases.

strohel replied:
Certainly not a concern in our simple and fun programs, but in tonari code we've learned that we should prefer returning an error (even if it's non-recoverable and bubbles all the way to `main()`) to panicking. (internal ref https://github.com/tonarino/portal/issues/2465#issuecomment-1523061135)
- panicking causes by default stack unwinding, but unwinding doesn't seem to play well with FFI, i.e. calling into CUDA, X11 etc (might just be that the FFI wrappers we use are not safe w.r.t. unwinding)
- destructors run during stack unwinding, and if they do panic (they cannot return a Result), such a double-panic causes an abort (skips any further destructors, i.e. could leave buffers not flushed etc)
  - Destructors will also run when returning an Err and they may panic as well, but you get "one more try" this way

So if we want to be defensive, we'd convert both cases to return an Error as you say.

skywhale replied:
Thank you :) I went lazy this time and converted this to `.expect()` 🙇

goodhoko replied:
@strohel Thanks for sharing these lessons from the wild (i.e. production). It sounds pragmatic but aren't you then paying for this

> even if it's non-recoverable and bubbles all the way to main()

with Options and Results everywhere all the way to main? Not that it's a big price to pay.

strohel replied:
> It sounds pragmatic but aren't you then paying for this with Options and Results everywhere all the way to main? Not that it's a big price to pay.

Yes. Though with actor framework as a middleman this isn't that striking: the Actor::handle() method is already fallible and System::run() is as well.

goodhoko commented on crates/dtrace_probe/src/main.rs:
```diff
@@ -28,19 +44,22 @@ fn main() -> Result<()> {
 
-    loop {
+    while running.load(Ordering::SeqCst) {
+        let mut message = EventMessage::new();
```
I think this can't compile - I guess you meant
```suggestion
        let mut message = EventMessage::default();
```

goodhoko reviewed (approved):
This looks good!

I tried to run this against the whole system (without pid) and it completely froze my laptop in a weird way. I'll do more testing but let's merge this.

skywhale committed:
Proper ctrl-c handling and optional process ID.

skywhale committed:
cargo fmt

skywhale committed:
cargo fmt with +nightly

skywhale commented:
I'll also go ahead and merge this as I have an approval. I addressed all of the comments -- I'm sorry I made a mistake during rebasing and all the changes were squashed into a single commit 🙇

skywhale commented:
@goodhoko my PC also had a couple of freezing moments 😢  I will dig deeper to see if I can stabilize it more.
